File: sources/metil_collision/metil_collision_uncollide/metil_collision_uncollide_circular.m
#include <metil_collision/metil_collision_uncollide/metil_collision_uncollide_circular.h>

#include <metil_object/metil_object.h>
#include <metil_player/metil_player.h>

#include <math_c_vector.h>
#include <math_c_vector_scale.h>
#include <math_c_vector_uncollide.h>

#include <math_c_maximum.h>
#include <math_c_power.h>
#include <math_c_square_root.h>

unsigned char metil_collision_object_uncollide_circular_xz(
  struct math_c_vector3_float* position_collidable,
  struct math_c_vector3_float* size_collidable,
  struct metil_object* metil_object_collider
) {
  return math_c_vector3_float_uncollide_circular_xz(
    position_collidable,
    size_collidable,
    &metil_object_collider->position,
    &metil_object_collider->mesh.size
  );
}

unsigned char metil_collision_object_uncollide_circular_distance_xz(
  struct math_c_vector3_float* position_collidable,
  struct metil_object* metil_object_collider,
  float distance_collision_minimum
) {
  return math_c_vector3_float_uncollide_circular_distance_xz(
    position_collidable,
    &metil_object_collider->position,
    distance_collision_minimum
  );
}

unsigned char metil_collision_player_uncollide_circular_xz(
  struct math_c_vector3_float* position_collidable,
  struct math_c_vector3_float* size_collidable,
  struct metil_player* metil_player_collider
) {
  return math_c_vector3_float_uncollide_circular_xz(
    position_collidable,
    size_collidable,
    &metil_player_collider->position,
    &metil_player_collider->size
  );
}

unsigned char metil_collision_player_uncollide_circular_distance_xz(
  struct math_c_vector3_float* position_collidable,
  struct metil_player* metil_player_collider,
  float distance_collision_minimum
) {
  return math_c_vector3_float_uncollide_circular_distance_xz(
    position_collidable,
    &metil_player_collider->position,
    distance_collision_minimum
  );
}

unsigned char metil_collision_object_object_uncollide_circular_xz(
  struct metil_object* metil_object_collidable,
  struct metil_object* metil_object_collider
) {
  return math_c_vector3_float_uncollide_circular_xz(
    &metil_object_collidable->position,
    &metil_object_collidable->mesh.size,
    &metil_object_collider->position,
    &metil_object_collider->mesh.size
  );
}

unsigned char metil_collision_object_object_uncollide_circular_distance_xz(
  struct metil_object* metil_object_collidable,
  struct metil_object* metil_object_collider,
  float distance_collision_minimum
) {
  return math_c_vector3_float_uncollide_circular_distance_xz(
    &metil_object_collidable->position,
    &metil_object_collider->position,
    distance_collision_minimum
  );
}

unsigned char metil_collision_player_object_uncollide_circular_xz(
  struct metil_object* metil_object_collidable,
  struct metil_player* metil_player_collider
) {
  return math_c_vector3_float_uncollide_circular_xz(
    &metil_object_collidable->position,
    &metil_object_collidable->mesh.size,
    &metil_player_collider->position,
    &metil_player_collider->size
  );
}

unsigned char metil_collision_player_object_uncollide_circular_distance_xz(
  struct metil_object* metil_object_collidable,
  struct metil_player* metil_player_collider,
  float distance_collision_minimum
) {
  return math_c_vector3_float_uncollide_circular_distance_xz(
    &metil_object_collidable->position,
    &metil_player_collider->position,
    distance_collision_minimum
  );
}

unsigned char metil_collision_object_player_uncollide_circular_xz(
  struct metil_player* metil_player_collidable,
  struct metil_object* metil_object_collider
) {
  return math_c_vector3_float_uncollide_circular_xz(
    &metil_player_collidable->position,
    &metil_player_collidable->size,
    &metil_object_collider->position,
    &metil_object_collider->mesh.size
  );
}

unsigned char metil_collision_object_player_uncollide_circular_distance_xz(
  struct metil_player* metil_player_collidable,
  struct metil_object* metil_object_collider,
  float distance_collision_minimum
) {
  return math_c_vector3_float_uncollide_circular_distance_xz(
    &metil_player_collidable->position,
    &metil_object_collider->position,
    distance_collision_minimum
  );
}

unsigned char metil_collision_vector_object_uncollide_circular_xz(
  struct metil_object* metil_object_collidable,
  struct math_c_vector3_float* position_collider,
  struct math_c_vector3_float* size_collider
) {
  return math_c_vector3_float_uncollide_circular_xz(
    &metil_object_collidable->position,
    &metil_object_collidable->mesh.size,
    position_collider,
    size_collider
  );
}

unsigned char metil_collision_vector_object_uncollide_circular_distance_xz(
  struct metil_object* metil_object_collidable,
  struct math_c_vector3_float* position_collider,
  float distance_collision_minimum
) {
  return math_c_vector3_float_uncollide_circular_distance_xz(
    &metil_object_collidable->position,
    position_collider,
    distance_collision_minimum
  );
}

unsigned char metil_collision_vector_player_uncollide_circular_xz(
  struct metil_player* metil_player_collidable,
  struct math_c_vector3_float* position_collider,
  struct math_c_vector3_float* size_collider
) {
  return math_c_vector3_float_uncollide_circular_xz(
    &metil_player_collidable->position,
    &metil_player_collidable->size,
    position_collider,
    size_collider
  );
}

unsigned char metil_collision_vector_player_uncollide_circular_distance_xz(
  struct metil_player* metil_player_collidable,
  struct math_c_vector3_float* position_collider,
  float distance_collision_minimum
) {
  return math_c_vector3_float_uncollide_circular_distance_xz(
    &metil_player_collidable->position,
    position_collider,
    distance_collision_minimum
  );
}

File: sources/metil_collision/metil_collision_uncollide/metil_collision_uncollide_circular.c
#include <metil_collision/metil_collision_uncollide/metil_collision_uncollide_circular.h>

#include <clic3_vector.h>

#include <math_c_maximum.h>
#include <math_c_power.h>
#include <math_c_square_root.h>

void metil_collision_uncollide_circular_xz(
  struct clic3_vector3_float* position_collidable,
  struct clic3_vector3_float* size_collidable,
  struct clic3_vector3_float* position_collider,
  struct clic3_vector3_float* size_collider
) {  
  float distance_collision_minimum = (
    math_c_maximum_float(
      size_collidable->x,
      size_collidable->z
    ) /
    2.0f +
    math_c_maximum_float(
      size_collider->x,
      size_collider->z
    ) /
    2.0f
  );

  metil_collision_uncollide_circular_distance_xz(
    position_collidable,
    position_collider,
    distance_collision_minimum
  );
}

void metil_collision_uncollide_circular_distance_xz(
  struct clic3_vector3_float* position_collidable,
  struct clic3_vector3_float* position_collider,
  float distance_collision_minimum
) {
  struct clic3_vector2_float distances_collidable_collider = {
    .x = (
      position_collidable->x -
      position_collider->x
    ),
    .y = (
      position_collidable->z -
      position_collider->z
    )
  };

  float distance_total_collidable_collider = (
    math_c_square_root(
      math_c_power_float(
        distances_collidable_collider.x,
        2.0f
      ) +
      math_c_power_float(
        distances_collidable_collider.y,
        2.0f
      )
    )
  );

  if (
    distance_total_collidable_collider >= distance_collision_minimum
  ) {
    return;
  }

  struct clic3_vector2_float distance_percentages_collidable_collider;

  if (
    distance_total_collidable_collider == 0.0f
  ) {
    distance_percentages_collidable_collider.x = (
      0.5f
    );

    distance_percentages_collidable_collider.y = (
      0.5f
    );
  } else {
    distance_percentages_collidable_collider.x = (
      distances_collidable_collider.x /
      distance_total_collidable_collider
    );

    distance_percentages_collidable_collider.y = (
      distances_collidable_collider.y /
      distance_total_collidable_collider
    );
  }

  float distance_difference = (
    distance_collision_minimum -
    distance_total_collidable_collider
  );

  position_collider->x = (
    position_collider->x -
    distance_percentages_collidable_collider.x *
    distance_difference
  );

  position_collider->z = (
    position_collider->z -
    distance_percentages_collidable_collider.y *
    distance_difference
  );
}
